### PR TITLE
fix shared library build with llvm-19

### PR DESCRIPTION
**Pull request by yamt**

### CODE REVIEW

**Comment by TerrorJack** [comment]
I'm seeing the following error when integrating it into a `wasi-sdk` build:

```
 "/workspace/wasi-sdk/build/install/bin/wasm-ld" -m wasm32 -L/workspace/wasi-sdk/build/sysroot/install/share/wasi-sysroot/lib/wasm32-wasi /workspace/wasi-sdk/build/sysroot/install/share/wasi-sysroot/lib/wasm32-wasi/crt1-reactor.o --entry _initialize -shared --whole-archive build/wasm32-wasi/libdl.so.a --no-whole-archive --allow-undefined-file=linker-provided-symbols.txt -lc /workspace/wasi-sdk/build/install/lib/clang/19/lib/wasm32-unknown-wasi/libclang_rt.builtins.a -o /workspace/wasi-sdk/build/sysroot/install/share/wasi-sysroot/lib/wasm32-wasi/libdl.so
wasm-ld: error: cannot open /workspace/wasi-sdk/build/install/lib/clang/19/lib/wasm32-unknown-wasi/libclang_rt.builtins.a: No such file or directory
wasm-ld: error: cannot open /workspace/wasi-sdk/build/install/lib/clang/19/lib/wasm32-unknown-wasi/libclang_rt.builtins.a: No such file or directory
wasm-ld: error: cannot open /workspace/wasi-sdk/build/install/lib/clang/19/lib/wasm32-unknown-wasi/libclang_rt.builtins.a: No such file or directory
```

Does this ring a bell to you?

**Comment by yamt** [comment]
> I'm seeing the following error when integrating it into a `wasi-sdk` build:
> 
> ```
>  "/workspace/wasi-sdk/build/install/bin/wasm-ld" -m wasm32 -L/workspace/wasi-sdk/build/sysroot/install/share/wasi-sysroot/lib/wasm32-wasi /workspace/wasi-sdk/build/sysroot/install/share/wasi-sysroot/lib/wasm32-wasi/crt1-reactor.o --entry _initialize -shared --whole-archive build/wasm32-wasi/libdl.so.a --no-whole-archive --allow-undefined-file=linker-provided-symbols.txt -lc /workspace/wasi-sdk/build/install/lib/clang/19/lib/wasm32-unknown-wasi/libclang_rt.builtins.a -o /workspace/wasi-sdk/build/sysroot/install/share/wasi-sysroot/lib/wasm32-wasi/libdl.so
> wasm-ld: error: cannot open /workspace/wasi-sdk/build/install/lib/clang/19/lib/wasm32-unknown-wasi/libclang_rt.builtins.a: No such file or directory
> wasm-ld: error: cannot open /workspace/wasi-sdk/build/install/lib/clang/19/lib/wasm32-unknown-wasi/libclang_rt.builtins.a: No such file or directory
> wasm-ld: error: cannot open /workspace/wasi-sdk/build/install/lib/clang/19/lib/wasm32-unknown-wasi/libclang_rt.builtins.a: No such file or directory
> ```
> 
> In the build directory, `compiler-rt` seems to be present in the following paths:
> 
> ```
> /workspace/wasi-sdk/build/sysroot/install/wasi-resource-dir/lib/wasi/libclang_rt.builtins-wasm32.a
> /workspace/wasi-sdk/build/sysroot/install/wasi-resource-dir/lib/wasip1/libclang_rt.builtins-wasm32.a
> /workspace/wasi-sdk/build/sysroot/install/wasi-resource-dir/lib/wasip2/libclang_rt.builtins-wasm32.a
> ```
> 
> I wonder why the path mismatch didn't happen before this patch though..

i pushed a commit which might fix it.

**Comment by TerrorJack** [comment]
Thanks, I've verified it fixes wasi-sdk build.

**Review by TerrorJack — APPROVED** [review]

File: linker-provided-symbols.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+__heap_base
+__heap_end
+__c_longjmp
```
Comment (sbc100) [inline linker-provided-symbols.txt:3]: Can you open a wasm-ld bug regarding this?   I think these should not be necessary. 

Also, perhaps add a comment here (or in the Makefile) with a link the bug so we know its temporary.

Comment (yamt) [reply linker-provided-symbols.txt:3]: done

Comment (yamt) [reply linker-provided-symbols.txt:3]: https://github.com/llvm/llvm-project/issues/103592

File: libc-top-half/musl/src/signal/psignal.c [hunk 2]
```diff
@@ -3,6 +3,11 @@
 #include <signal.h>
 #include <errno.h>
 
+/* undef the macro to use the standard stderr instead of __stderr_FILE
```
[the remark this answers is not in the document]
Comment (yamt) [reply libc-top-half/musl/src/signal/psignal.c:6]: sorry, i failed to parse your statement. can you rephrase?

[the remark this answers is not in the document]
Comment (sbc100) [reply libc-top-half/musl/src/signal/psignal.c:6]: Since you are changing musl sources and you wrap your changes `#ifdef __wasilibc_unmodified_upstream`

[the remark this answers is not in the document]
Comment (yamt) [reply libc-top-half/musl/src/signal/psignal.c:7]: oh i got it. done.

**Review by sbc100 — APPROVED** [review]
